src/router_discovery.erl

```diff
@@ -0,0 +1,135 @@
+                router_device:nwk_s_key(Device),
+                router_device:app_s_key(Device),
+                %% TODO: Fcnt should go up here
+                1,
+                #{body => <<1:8, TxnID/binary>>}
+            ),
+            send(PubKeyBin, DeviceID, Packet, Sig)
+    end.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+
+-spec send(
+    PubKeyBin :: libp2p_crypto:pubkey_bin(),
+    DeviceID :: binary(),
+    TxnID :: binary(),
+    Sig :: binary()
+) -> ok.
+send(PubKeyBin, DeviceID, TxnID, Sig) ->
+    case router_discovery_handler:dial(PubKeyBin) of
+        {error, _Reason} ->
+            lager:info("failed to dial hotpost ~p: ~p", [
+                blockchain_utils:addr2name(PubKeyBin),
+                _Reason
+            ]);
+        {ok, Stream} ->
+            lager:info(
+                "Sent discovery request to hotspot ~p (device_id=~p txn_id=~p)",
+                [
+                    blockchain_utils:addr2name(PubKeyBin),
+                    DeviceID,
+                    TxnID
+                ]
+            ),
+            BinMsg = discovery_pb:encode_msg(#discovery_start_pb{
+                hotspot = PubKeyBin,
+                packet = TxnID,
+                signature = Sig
+            }),
+            ok = router_discovery_handler:send(Stream, BinMsg)
+    end.
+
+-spec verify_signature(
+    Hostpost :: binary(),
+    HostpostPubKeyBin :: libp2p_crypto:pubkey_bin(),
+    Sig :: binary()
+) -> boolean().
+verify_signature(Hostpost, HostpostPubKeyBin, Sig) ->
+    case get_hotspot_owner(HostpostPubKeyBin) of
+        {error, _Reason} ->
+            lager:info("failed to find owner for hotspot ~p: ~p", [
+                {Hostpost, HostpostPubKeyBin},
+                _Reason
+            ]),
+            false;
+        {ok, OwnerPubKeyBin} ->
+            libp2p_crypto:verify(
+                <<Hostpost/binary>>,
+                base64:decode(Sig),
+                libp2p_crypto:bin_to_pubkey(OwnerPubKeyBin)
+            )
+    end.
+
+-spec get_hotspot_owner(PubKeyBin :: libp2p_crypto:pubkey_bin()) ->
+    {ok, libp2p_crypto:pubkey_bin()} | {error, any()}.
+get_hotspot_owner(PubKeyBin) ->
+    Chain = blockchain_worker:blockchain(),
+    Ledger = blockchain:ledger(Chain),
+    blockchain_ledger_v1:find_gateway_owner(PubKeyBin, Ledger).
+
+-spec frame_payload(
+    MType :: non_neg_integer(),
+    DevAddr :: binary(),
+    NwkSessionKey :: binary(),
+    AppSessionKey :: binary(),
+    FCnt :: non_neg_integer(),
+    Options :: map()
+) -> binary().
+frame_payload(MType, DevAddr, NwkSessionKey, AppSessionKey, FCnt, Options) ->
+    OptionsBoolToBit = fun(Key) ->
+        case maps:get(Key, Options, false) of
+            true -> 1;
+            false -> 0
+        end
+    end,
+    MHDRRFU = 0,
+    Major = 0,
+    ADR = OptionsBoolToBit(wants_adr),
```

Comment:
I can't think of any reason why discovery packets should set ADR bit. We probably w'ont send enough packets to build a full ADR history, but if we did, downlinks will be wasted since they won't be received by anything.

Reply:
Yeah I ll remove that this was just a function taken from tests